configuration.md
# Configuration options #

## Please note, this file does not contain all the options ##

**check http://code.google.com/p/gross/source/browse/trunk/doc/examples/grossd.conf for the complete list**

  * `dnsbl` option is the address of dns blacklist server. There can be multiple `dnsbl` options in the configuration file and each one will be queried in parallel. As soon as one of the blacklist servers recognizes the senders address, the mail will be greylisted. If none of the blacklist servers contain the senders address, then the sender is **trusted**. These values are unused if _grossd_ is configured with `--disable-dnsbl` flag.
  * `host` option is the address of the interface _grossd_ binds itself, default is `localhost`
  * `port` is the port number _grossd_ binds itself, default is `5525`
  * `stat_interval` is the number of _seconds_ between statistics logging
  * `stat_level` is the name of the requested statistic. There can be multiple `stat_level` options in the configuration file (Using both `none` and `full` is undefined). Default is `none`. Valid options are currently:
    * `full` = _grossd_ sends all possible statistics
    * `none` = _grossd_ will not send any statistics
    * `status` = _grossd_ will send basic statistics from the `stat_interval` period
    * `since_startup` = _grossd_ will send basic statistics since the server startup
    * `delay` = _grossd_ will send statistics about query delays
  * `filter_bits` option determines the size of the bloom filter. The size will be 2^filter\_bits, default is `22`.
  * `number_buffers` is the number of bloom filters used in the ring queue, default is `8`.
  * `rotate_interval` is the number of _seconds_ between bloom filter rotation. An individual entry will be removed from the filter after `rotate_interval` times the `number_buffers` _seconds_ at the earliest. Default is `3600`
  * `sync_peer` is the address of the peer used in clustered mode
  * `sync_listen` is the address of the interface to bind to listen for peer communication, default is the value of `host`.
  * `sync_port` is the port number to listen to and connect to in peer communication, default is `5524`
  * `statefile` is path to file that is used to store the state (filters etc.) of the _grossd_ instance
  * `log_method` `syslog` is currently the only method implemented, it is the default.
  * `log_level` default is `info`, the other valid options are `debug`, `notice`, `error` and `warning`
  * `syslog_facility` is the facility _syslog_ sends log messages with.
  * `update` is the way _grossd_ updates its filters. If `update = grey`, filters are only updated if response is `STATUS_GREY`. `update = always` makes _grossd_ to update filters on every query. Default is `grey`.
  * `grey_delay` is the number of _seconds_ between accepting the second sending attempt after greylisting. The default value is 10, increasing the delay causes the amortized memory requirement to grow linearly.
  * `grey_mask` is the mask for grossd to use when matching client\_ip against the database. You may set it to 24 to treat  addresses like a.b.c.d as a.b.c.0.

NewConfigFileFormat.md
# Introduction #

I found a solid candidate for the config file parser library for Gross. Its name is [Confuse](http://www.nongrnu.org/confuse). It's fully featured and seems quite mature and clean. And it has a compatible licence so we can include it in the source tree.

# Config file prototype #

```
# This is subject to chanage

global {
    host = ${HOST:-localhost} # environment variables!
    statefile = /var/db/grossd.state
    pid {
        file = /var/db/grossd.pid
        check = true
    }
    logging syslog {
         level = notice
         facility = mail
    }
    logging file {
        level = info
        path = /var/log/grossd.log
    }
}

protocol sjsms {
    host = localhost # if differs from global option
    port = 1111
}

protocol postfix {}

protocol milter {
    listen = inet:5523@localhost
}

check spamhaus {
    type = dnsbl
    zone = zen.spamhaus.org
    in 127.0.0.4/30 {
        weight = 4
    }
    in 127.0.0.2/32 {
        weight = 8
    }
    weight = 1
}

check spamcop {
    type = dnsbl
    zone = bl.spamcop.net
    weight = 2
    delay = 100 ms
}

check sorbs {
    type = dnsbl
    zone = dnsbl.sorbs.net
   delay = 200 ms
}

check bondedsender {
    type = dnswl
    zone = query.bondedsender.org
}

check blocker {
   type = blocker
   host = localhost
   port = 4466
   weight = 1
}

policy {
    block {
        threshold = 9
        reason = "Bad reputation"
        replycode = 502
    }
    greylist {
        threshold = 1
        delay = 10s
        reason = "Please, try again later"
        replycode = 452
        mask = 24
    }
    timelimit = 5s
}

```

NEWS
Version svn

* New check: reverse. It tries to find a name for client_ip
  and then check if that name points back to client_ip.
* New check: helo. It checks if helo_name matches client_ip.
* New check: spf. Depends on libspf2, compile time option.
* New configure option 'grey_tuple' for looser greylist.
* do not check /etc/hosts in dnsbl check
* -u command line option to run grossd with a differnet uid

Issues fixed:
#71: grossd dies under Linux
#72: freebsd port broken

Version 1.0.1

* Two new configuration options: postfix_response_grey and
  postfix_response_block. Now it's possible to configure
  responses grossd sends in reply to Postfix policy delegation
  client.
* Now grossd allows only the configured sync_peer to connect
  to sync server port.
* When grey_threshold = 0 is set grossd defaults to greylisting.
  Now it's possible to combine traditional greylisting with 
  blocking based on check results.

Issues fixed:
#65: Version number to log (RFE)
#66: Compiler warnings on with 1.0.0
#67: Setting grey_threshold = 0 is not intuitive
#68: grossd should accept sync connections only from the configured sync_peer
#69: Implement postfix_response_grey and postfix_response_block (RFE)
#70: Busyloop if no access to configured statefile 

Version 1.0.0

* Prevent grossd from running as root (euid == 0). You may 
  have to modify configuration and state file ownerships
  and/or permissions. Grossd tries to setuid() and setgid()
  to nobody if started by root.
* New match shortcut logic. Now grossd will shortcut after
  a filter match only if no checks are configured (traditional
  greylister) or block_threshold == 0. 

- changed filter_bits default back to 24, as stated in the docs.
- added grey_reason configuration file option. See block_reason.

Issues fixed:
#48: Need man pages
#59: grossd 0.9.3 leaks memory
#60: Grossd should do all dns queries as fully qualified
#61: Space Character in Mail From Crashing Gross
#62: Double free() with milter

Version 0.9.3

* A few important bug fixes to milter server
* HELOified all the rest of the code. There are still no checks
  that use helo info, but now grossd logs it and also gclient 
  can send it to the server. Check INSTALL details about 
  configuring SJSMS to send helo string over to grossd.
* New logfile format! An example log line: 

  Thu Apr 17 14:50:05 2008 #8505600: a=greylist d=3 w=5 c=127.0.0.2
  s=sender r=recipient h=N/A m=cbl.abuseat.org+1 m=bl.spamcop.net+2
  m=combined.njabl.org+1 m=dnsbl.sorbs.net+1

  where
  a = action
  d = delay
  w = total weight
  c = client ip
  s = sender
  r = recipient
  h = helo/ehlo
  m = matched check with possible weight

- A few important bug fixes to milter server
- behaviour change: grossd now exits after creating the
  statefile when invoked with -C command line option. This
  makes life easier for package maintainers in the future.
- Implemented pidfile creation and checking. New config option
  'pidfile' controls the behaviour. Setting the option will make
  grossd to write the process pid into the file specified by the
  full path. An optional parameter 'check' can be set to keep
  grossd from starting if the pidfile already exists.

  pidfile = /full/path[;check]

  Another way to control this is by command line options -p and -P.
  Both of the options will enable pidfile creation, and -P will
  check the existence of the file.

Issues fixed:
#58: grossd should write a pid file after daemonizing (RFE)

Version 0.9.2

- documented configuration options query_timelimit and
  block_reason. See doc/examples/grossd.conf
- new option: pool_maxthreads, you may need to change this
  if you get more than 100 queries per second and/or have
  slow dns servers
- some log messages got sent with wrong priorities

#56: Grossd skips all the checks if block_threshold not set
#57: sjsms: query_timelimit is one second off

Version 0.9.1

Issues fixed:
#55: --disable-dnsbl broken on 0.9

Version 0.9

* changed the port defaults to be more compatible with
  the community, grossd now defaults to ports 5525 (grossd),
  5524 (sync), and 5522 (status queries)
* preliminary support for MILTER protocol, check config file
* support for definitive checks: it's now possible to write 
  checks that cause grossd to block or pass (whitelisting)
* dnsbl check is now run time configurable
* two new check types: RHSBL and DNSWL - check config file
  for documentation and examples. DNSWL is first definitive
  check. 

- improved thread counting in thread_pool.c
- code cleaning
- default for grey_mask was 0, effectively discarding client_ip
  information from greylisting triplet. Changed it to 24.
- checks for c-ares version: there's an API change between
  1.4.x and 1.5.0

Issues fixed:
#40: new type of check - RHSBL (RFE)
#41: new type of check - DNSWL (RFE)
#46: Make dnsbl checking run time configurable (RFE)
#47: Grossd will not run any checks if configured with --disable-dnsbl
#54: Grossd dies without error msg in non-localhost setup (with linux)

Version 0.8.2

- updating mappings example in INSTALL
- logging less verbose (one line per query)

Issues fixed:
#39: blocker queries cause segmentation fault
#43: Lot's of open udp connections when using grosscheck

Version 0.8.1

A couple of bugs fixed.

Issues fixed:
#35: grossd crashes when starting
#36: synchronization still loses one update message

Version 0.8.0

In short: Important changes from the previous version

* You must configure protocols in config file. This is done via
  'protocol' config file parameter. Either 'protocol = sjsms' or
  'protocol = postfix' or both.
* 'sync_host' and 'peer_host' has been deprecated. Use new
  'sync_listen' and 'sync_peer' options.


The longer list:
- Checks are now done in thread pools, which makes it easier to
  develop new checks, like Sophos blocker or SPF. 
- grossd can log aggregate statistics about query results. Check
  'stat_type' and 'stat_interval' configuration parameters.

include/conf.h
/* $Id$ */

/*
 * Copyright (c) 2006,2007,2008
 *               Eino Tuominen <eino@utu.fi>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#ifndef CONF_H
#define CONF_H

#define DEFAULT_CONFIG	"update",		"grey", 	\
			"host",			"127.0.0.1",	\
			"port",			"5525",		\
			"sync_port",		"5524",		\
			"status_port",		"5522",		\
			"rotate_interval", 	"3600",		\
			"filter_bits",		"24",		\
			"number_buffers",	"8",            \
			"stat_interval",	"300",		\
			"postfix_response_grey","action=defer_if_permit %reason%", \
			"postfix_response_block","action=reject %reason%",	\
			"sjsms_response_grey",	"$X4.4.3|$N%reason%", \
			"sjsms_response_match",	"$Y", 		\
			"sjsms_response_trust",	"$Y",		\
			"sjsms_response_block", "$N%reason%",	\
			"log_method",		"syslog",	\
			"log_level",		"info",		\
			"stat_type",		"delay",	\
			"stat_type",		"status",	\
			"grey_mask",		"24",		\
			"grey_delay",		"10",           \
			"grey_tuple",		"user",		\
			"syslog_facility",	"mail",		\
			"blocker_port",		"4466",		\
			"blocker_weight",	"1",		\
			"block_threshold",	"0",		\
			"grey_threshold",	"1",		\
			"grey_reason",		"Please try again later", \
			"block_reason",		"Bad reputation", \
			"query_timelimit",	"5000",		\
			"pool_maxthreads",	"100"

#define MULTIVALUES	"dnsbl",	\
			"rhsbl",	\
			"dnswl",	\
			"check",	\
                        "stat_type",	\
			"protocol", 	\
			"log_method"

#define VALID_NAMES     "dnsbl",			\
			"rhsbl",			\
			"dnswl",			\
			"host",				\
			"port",				\
                        "filter_bits",			\
                        "rotate_interval",		\
                        "number_buffers",		\
                        "update",			\
                        "peer_name",			\
                        "statefile",			\
			"postfix_response_grey",	\
			"postfix_response_block",	\
			"sjsms_response_grey",		\
			"sjsms_response_match",		\
			"sjsms_response_trust",		\
			"sjsms_response_block",		\
                        "log_method",			\
                        "log_level",			\
			"grey_mask",			\
                        "grey_delay",               	\
                        "grey_tuple",               	\
			"check",			\
			"protocol",			\
                        "syslog_facility",		\
                        "sync_listen",			\
                        "sync_peer",			\
                        "sync_port",			\
                        "stat_interval",		\
                        "stat_type",			\
			"status",			\
			"blocker_host",			\
			"blocker_port",			\
			"blocker_weight",		\
			"query_timelimit",		\
			"block_threshold",		\
			"grey_threshold",		\
			"grey_reason",			\
			"block_reason",			\
			"milter_listen",		\
			"pidfile",			\
			"pool_maxthreads"

#define DEPRECATED_NAMES 	"syncport",		\
				"synchost",		\
				"peerport",		\
				"peerhost",		\
				"statushost",		\
				"statusport"


#define RUNTIME_MODIFIABLE	"grey_threshold",		\
				"block_threshold",		\
				"pool_maxthreads",		\
				"grey_mask",			\
				"grey_delay",			\
				"log_level",			\
				"check",			\
				"query_timelimit",		\
				"postfix_response_grey",        \
				"postfix_response_block",       \
				"sjsms_response_grey",          \
				"sjsms_response_match",         \
				"sjsms_response_trust",         \
				"sjsms_response_block",         \
				"grey_reason", 		        \
				"block_reason", 	        \
				"update"
/*
 * How many parameters a given keyword accepts, format is
 * "keyword", MIN, MAX. -1 as maximum means unlimited
 */
#define PARAMS	"dnsbl",	"0",	"1",	\
		"rhsbl",	"0",	"1",	\
		"pidfile",	"0",	"1"

typedef struct params_s
{
	const char *value;
	struct params_s *next;	/* linked list */
} params_t;

typedef struct configlist_s
{
	bool is_default;
	const char *name;
	const char *value;
	params_t *params;
	struct configlist_s *next;	/* linked list */
} configlist_t;

configlist_t *default_config(void);
int read_config(configlist_t **config, const char *filename);
const char *gconf(configlist_t *config, const char *name);

#endif /* CONF_H */
